issue while backup and restore from savepoint

This is regarding backup restore . We are not taking any LPV backup. While restore the old LPV won't be present. We are taking savepoint backup ,while restoring getting below exception from flink connector. 
Connector should take offset from savepoint (which is backed up) and process.

Flink version used **1.9.3**, pulsar-flink-connector_2.11-**2.4.28.3** and pulsar **2.6**

**Log:**
2021-04-05 08:32:05.894 [Async calls on Source: src5 -> (src5_monitoring, invalid_data_topic_monitoring, Sink: invalid_data_topic, String_to_JsonNode -> Timestamps/Watermarks) (7/8)] WARN  o.a.flink.streaming.connectors.pulsar.internal.PulsarFetcher  - **Failed to commit cursor to Pulsar.**
java.lang.RuntimeException: **Failed to commit cursor offset 13086:38:0:0** for persistent://public/default/src5-partition-0
                    at org.apache.flink.streaming.connectors.pulsar.internal.PulsarMetadataReader.commitCursorToOffset(PulsarMetadataReader.java:263)
                    at org.apache.flink.streaming.connectors.pulsar.internal.PulsarFetcher.doCommitOffsetToPulsar(PulsarFetcher.java:452)
                    at org.apache.flink.streaming.connectors.pulsar.internal.PulsarFetcher.commitOffsetToPulsar(PulsarFetcher.java:440)
                    at org.apache.flink.streaming.connectors.pulsar.FlinkPulsarSource.notifyCheckpointComplete(FlinkPulsarSource.java:742)
                    at org.apache.flink.streaming.api.operators.AbstractUdfStreamOperator.notifyCheckpointComplete(AbstractUdfStreamOperator.java:130)
                    at org.apache.flink.streaming.runtime.tasks.StreamTask.notifyCheckpointComplete(StreamTask.java:842)
                    at org.apache.flink.runtime.taskmanager.Task$2.run(Task.java:1214)
                    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
                    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
                    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
                    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
                    at java.lang.Thread.run(Thread.java:748)
Caused by: org.apache.pulsar.client.admin.PulsarAdminException: A MultiException has 1 exceptions.  They are:
1. java.lang.IllegalStateException: ServiceLocatorImpl(__HK2_Generated_128,128,356740187) has been shut down
                    at org.apache.pulsar.client.admin.internal.BaseResource.getApiException(BaseResource.java:228)
                    at org.apache.pulsar.client.admin.internal.TopicsImpl.resetCursor(TopicsImpl.java:864)
                    at org.apache.flink.streaming.connectors.pulsar.internal.PulsarMetadataReader.commitCursorToOffset(PulsarMetadataReader.java:253)
                    ... 11 common frames omitted
Caused by: org.apache.pulsar.shade.org.glassfish.hk2.api.MultiException: A MultiException has 1 exceptions.  They are:
1. java.lang.IllegalStateException: ServiceLocatorImpl(__HK2_Generated_128,128,356740187) has been shut down
                    at org.apache.pulsar.shade.org.jvnet.hk2.internal.FactoryCreator.getFactoryHandle(FactoryCreator.java:106)
                    at org.apache.pulsar.shade.org.jvnet.hk2.internal.FactoryCreator.dispose(FactoryCreator.java:173)
                    at org.apache.pulsar.shade.org.jvnet.hk2.internal.SystemDescriptor.dispose(SystemDescriptor.java:542)
                    at org.apache.pulsar.shade.org.glassfish.jersey.inject.hk2.RequestContext.lambda$findOrCreate$0(RequestContext.java:84)
                    at org.apache.pulsar.shade.org.glassfish.jersey.internal.inject.ForeignDescriptorImpl.dispose(ForeignDescriptorImpl.java:87)
                    at org.apache.pulsar.shade.org.glassfish.jersey.inject.hk2.Hk2RequestScope$Instance.remove(Hk2RequestScope.java:150)
                    at java.lang.Iterable.forEach(Iterable.java:75)
                    at org.apache.pulsar.shade.org.glassfish.jersey.inject.hk2.Hk2RequestScope$Instance.release(Hk2RequestScope.java:167)
                    at org.apache.pulsar.shade.org.glassfish.jersey.process.internal.RequestScope.release(RequestScope.java:270)
                    at org.apache.pulsar.shade.org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:416)
                    at org.apache.pulsar.shade.org.glassfish.jersey.client.JerseyInvocation.invoke(JerseyInvocation.java:765)
                    at org.apache.pulsar.shade.org.glassfish.jersey.client.JerseyInvocation$Builder.method(JerseyInvocation.java:456)
                    at org.apache.pulsar.shade.org.glassfish.jersey.client.JerseyInvocation$Builder.post(JerseyInvocation.java:357)
                    at org.apache.pulsar.client.admin.internal.TopicsImpl.resetCursor(TopicsImpl.java:861)
                    ... 12 common frames omitted
Caused by: java.lang.IllegalStateException: ServiceLocatorImpl(__HK2_Generated_128,128,356740187) has been shut down
                    at org.apache.pulsar.shade.org.jvnet.hk2.internal.ServiceLocatorImpl.checkState(ServiceLocatorImpl.java:2393)
                    at org.apache.pulsar.shade.org.jvnet.hk2.internal.ServiceLocatorImpl.getServiceHandleImpl(ServiceLocatorImpl.java:635)
                    at org.apache.pulsar.shade.org.jvnet.hk2.internal.ServiceLocatorImpl.getServiceHandle(ServiceLocatorImpl.java:628)
                    at org.apache.pulsar.shade.org.jvnet.hk2.internal.ServiceLocatorImpl.getServiceHandle(ServiceLocatorImpl.java:646)
                    at org.apache.pulsar.shade.org.jvnet.hk2.internal.FactoryCreator.getFactoryHandle(FactoryCreator.java:103)
                    ... 25 common frames omitted

--- a/pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/internal/PulsarMetadataReader.java
+++ b/pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/internal/PulsarMetadataReader.java
@@ -32,12 +32,14 @@ import org.apache.pulsar.common.schema.SchemaInfo;
 import org.apache.pulsar.shade.com.google.common.collect.Iterables;
 import org.apache.pulsar.shade.com.google.common.collect.Sets;
 
+import java.text.MessageFormat;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Map;
+import java.util.Optional;
 import java.util.Set;
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
@@ -406,6 +408,24 @@ public class PulsarMetadataReader implements AutoCloseable {
             return this.admin.topics().getLastMessageId(topic);
         } catch (PulsarAdminException e) {
             throw new RuntimeException(e);
+        }
+    }
+
+    public boolean checkCursorAvailable(String topic, MessageIdImpl startMessageId) {
+        try {
+            PersistentTopicInternalStats stats = this.admin.topics().getInternalStats(topic);
+            long ledgerId = startMessageId.getLedgerId();
+            // Pulsar's ledger is out of order and cannot be compared by obtaining the last ledger.
+            // Therefore, it is a safer way to check whether the current ledger exists.
+            final Optional<PersistentTopicInternalStats.LedgerInfo> ledgerInfo = stats.ledgers.stream()
+                .filter(l -> l.ledgerId == ledgerId)
+                .findAny();
+            return !ledgerInfo.filter(info -> startMessageId.getEntryId() > info.entries).isPresent();
+        } catch (Exception e) {
+            String message = MessageFormat.format(
+                "valid Cursor fail topic [{0}], messageId [{2}]",
+                topic, startMessageId.toString());
+            throw new RuntimeException(message, e);
         }
     }
 

--- a/pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/internal/ReaderThread.java
+++ b/pulsar-flink-connector/src/main/java/org/apache/flink/streaming/connectors/pulsar/internal/ReaderThread.java
@@ -140,23 +140,21 @@ public class ReaderThread<T> extends Thread {
         if (!startMessageId.equals(MessageId.earliest)
                 && !startMessageId.equals(MessageId.latest)
                 && ((MessageIdImpl) startMessageId).getEntryId() != -1) {
-            MessageIdImpl lastMessageId = (MessageIdImpl) this.owner.getMetaDataReader().getLastMessageId(reader.getTopic());
+            final PulsarMetadataReader metaDataReader = this.owner.getMetaDataReader();
+            MessageIdImpl lastMessageId = (MessageIdImpl) metaDataReader.getLastMessageId(reader.getTopic());
             if (!messageIdRoughEquals(startMessageId, lastMessageId) && !reader.hasMessageAvailable()) {
                 MessageIdImpl startMsgIdImpl = (MessageIdImpl) startMessageId;
-                long startMsgLedgerId = startMsgIdImpl.getLedgerId();
-                long startMsgEntryId = startMsgIdImpl.getEntryId();
-
                 // startMessageId is bigger than lastMessageId
-                if (startMsgLedgerId > lastMessageId.getLedgerId()
-                    || (startMsgLedgerId == lastMessageId.getLedgerId() && lastMessageId.getEntryId() != -1
-                        && startMsgEntryId > lastMessageId.getEntryId())) {
-                    log.error("the start message id is beyond the last commit message id, with topic:{}, " +
-                            "start msgId:{}, last msgId:{}", reader.getTopic(), startMessageId, lastMessageId);
+                if (!metaDataReader.checkCursorAvailable(reader.getTopic(), startMsgIdImpl)) {
+                    if (failOnDataLoss) {
+                        log.error("the start message id is beyond the last commit message id, with topic:{}", reader.getTopic());
+                        throw new RuntimeException("start message id beyond the last commit");
+                    } else {
+                        log.info("reset message to valid offset {}", startMessageId);
+                        metaDataReader.resetCursor(reader.getTopic(), startMessageId);
+                    }
                 }
-                log.warn("reset message to valid offset");
-                this.owner.getMetaDataReader().resetCursor(reader.getTopic(), startMessageId);
-            } else {
-                failOnDataLoss = false;
+
             }
             while (currentMessage == null && running) {
                 currentMessage = reader.readNext(pollTimeoutMs, TimeUnit.MILLISECONDS);

